File: backend/src/services/measurement.py
"""脊柱测量算法（Python 版本）- 与 Next.js 完全对齐"""
import math
from typing import List, Dict, Tuple, Optional
from ..schemas.screening import Landmark


class SpineMeasurement:
    """脊柱测量工具"""

    # MediaPipe 关键点索引
    LEFT_SHOULDER = 11
    RIGHT_SHOULDER = 12
    LEFT_HIP = 23
    RIGHT_HIP = 24
    LEFT_KNEE = 25
    RIGHT_KNEE = 26

    SEVERITY_SUMMARY = {
        "balanced": "脊柱形态正常，肩骨盆基本平衡。建议保持良好站姿和运动习惯。",
        "attention": "检测到轻度不对称，建议 1-2 个月后复测。可加强核心肌群训练，注意书包重量和坐姿。",
        "alert": "检测到明显不对称，建议尽快到医院脊柱外科就诊，进行 X 光检查以确认 Cobb 角。"
    }

    @staticmethod
    def pick_landmark(landmarks: List[Landmark], index: int) -> Optional[Landmark]:
        """选择有效的关键点"""
        if index >= len(landmarks):
            return None
        lm = landmarks[index]
        if lm.visibility is not None and lm.visibility < 0.3:
            return None
        return lm

    @staticmethod
    def distance(p1: Landmark, p2: Landmark) -> float:
        """计算两点之间的欧氏距离"""
        return math.sqrt((p2.x - p1.x) ** 2 + (p2.y - p1.y) ** 2)

    @staticmethod
    def midpoint(a: Landmark, b: Landmark) -> Dict:
        """获取两点的中点"""
        return {
            "x": (a.x + b.x) / 2,
            "y": (a.y + b.y) / 2,
            "visibility": min(a.visibility or 1, b.visibility or 1)
        }

    @classmethod
    def calc_spinal_curvature(
        cls,
        shoulder_mid: Dict,
        hip_mid: Dict,
        l_shoulder: Landmark,
        r_shoulder: Landmark,
        l_hip: Landmark,
        r_hip: Landmark
    ) -> float:
        """计算脊柱曲线角度（类 Cobb 角）"""
        # 方法1：理想垂直中线的偏离
        ideal_mid_x = (shoulder_mid["x"] + hip_mid["x"]) / 2
        torso_length = math.sqrt(
            (shoulder_mid["x"] - hip_mid["x"]) ** 2 +
            (shoulder_mid["y"] - hip_mid["y"]) ** 2
        )
        max_deviation = max(
            abs(shoulder_mid["x"] - ideal_mid_x),
            abs(hip_mid["x"] - ideal_mid_x)
        )
        deviation_angle = math.degrees(math.atan2(max_deviation, torso_length))

        # 方法2：肩膀和髋部的高度不对称性
        shoulder_asymmetry_angle = math.degrees(
            math.atan2(abs(l_shoulder.y - r_shoulder.y), 1.0)
        )
        hip_asymmetry_angle = math.degrees(
            math.atan2(abs(l_hip.y - r_hip.y), 1.0)
        )

        # 方法3：躯干侧移
        torso_lateral_shift = abs(shoulder_mid["x"] - hip_mid["x"])
        lateral_shift_angle = math.degrees(math.atan2(torso_lateral_shift, torso_length))

        # 综合评分（权重与 Next.js 一致）
        total_angle = (
            shoulder_asymmetry_angle * 2.0 +
            hip_asymmetry_angle * 1.5 +
            lateral_shift_angle * 1.5 +
            deviation_angle * 1.0
        )

        return total_angle

    @classmethod
    def assess_posture_quality(
        cls,
        l_shoulder: Landmark,
        r_shoulder: Landmark,
        l_hip: Landmark,
        r_hip: Landmark,
        l_knee: Optional[Landmark],
        r_knee: Optional[Landmark]
    ) -> float:
        """评估姿势质量（0-1）"""
        confidence = 1.0

        # 检查关键点可见性
        visibilities = [
            l_shoulder.visibility or 1,
            r_shoulder.visibility or 1,
            l_hip.visibility or 1,
            r_hip.visibility or 1
        ]
        avg_visibility = sum(visibilities) / len(visibilities)
        if avg_visibility < 0.7:
            confidence *= 0.7

        # 检查身体是否正对相机
        shoulder_width = abs(l_shoulder.x - r_shoulder.x)
        hip_width = abs(l_hip.x - r_hip.x)
        if hip_width > 0:
            width_ratio = shoulder_width / hip_width
            if width_ratio < 1.0 or width_ratio > 2.0:
                confidence *= 0.8

        # 检查双腿是否并拢
        if l_knee and r_knee:
            knee_distance = abs(l_knee.x - r_knee.x)
            shoulder_distance = abs(l_shoulder.x - r_shoulder.x)
            if knee_distance > shoulder_distance * 0.8:
                confidence *= 0.9

        return confidence

    @classmethod
    def classify_severity(
        cls,
        curvature: float,
        shoulder_angle: float,
        pelvis_angle: float,
        confidence: float
    ) -> str:
        """分类严重程度"""
        if confidence < 0.6:
            return "attention"

        max_angle = max(curvature, abs(shoulder_angle), abs(pelvis_angle))

        if max_angle < 8:
            return "balanced"
        if max_angle < 15:
            return "attention"
        return "alert"

    @classmethod
    def calculate_metrics(
        cls,
        landmarks: List[Landmark],
        image_width: int = 640,
        image_height: int = 480
    ) -> Dict:
        """
        计算脊柱测量指标（与 Next.js 版本完全对齐）

        Args:
            landmarks: 33个关键点列表
            image_width: 图像宽度
            image_height: 图像高度

        Returns:
            包含所有测量指标的字典
        """
        if len(landmarks) < 33:
            raise ValueError(f"需要33个关键点，但只收到 {len(landmarks)} 个")

        # 获取关键点
        l_shoulder = cls.pick_landmark(landmarks, cls.LEFT_SHOULDER)
        r_shoulder = cls.pick_landmark(landmarks, cls.RIGHT_SHOULDER)
        l_hip = cls.pick_landmark(landmarks, cls.LEFT_HIP)
        r_hip = cls.pick_landmark(landmarks, cls.RIGHT_HIP)
        l_knee = cls.pick_landmark(landmarks, cls.LEFT_KNEE)
        r_knee = cls.pick_landmark(landmarks, cls.RIGHT_KNEE)

        if not all([l_shoulder, r_shoulder, l_hip, r_hip]):
            raise ValueError("关键关节点不可见，请确保背部完整入镜")

        # 计算中点
        shoulder_mid = cls.midpoint(l_shoulder, r_shoulder)
        hip_mid = cls.midpoint(l_hip, r_hip)

        # 评估姿势质量
        posture_confidence = cls.assess_posture_quality(
            l_shoulder, r_shoulder, l_hip, r_hip, l_knee, r_knee
        )

        # 计算脊柱曲线角度
        spinal_curvature_deg = cls.calc_spinal_curvature(
            shoulder_mid, hip_mid, l_shoulder, r_shoulder, l_hip, r_hip
        )

        # 计算传统指标
        shoulder_height_diff_px = abs(l_shoulder.y - r_shoulder.y) * image_height
        shoulder_slope_deg = abs(math.degrees(
            math.atan2(r_shoulder.y - l_shoulder.y, r_shoulder.x - l_shoulder.x)
        ))
        pelvis_tilt_deg = abs(math.degrees(
            math.atan2(r_hip.y - l_hip.y, r_hip.x - l_hip.x)
        ))

        # 分类严重程度
        severity = cls.classify_severity(
            spinal_curvature_deg, shoulder_slope_deg, pelvis_tilt_deg, posture_confidence
        )

        return {
            "shoulderHeightDiffPx": round(shoulder_height_diff_px, 1),
            "shoulderSlopeDeg": round(shoulder_slope_deg, 2),
            "pelvisTiltDeg": round(pelvis_tilt_deg, 2),
            "spinalCurvatureDeg": round(spinal_curvature_deg, 2),
            "postureConfidence": round(posture_confidence, 2),
            "severity": severity,
            "summary": cls.SEVERITY_SUMMARY[severity],
        }

File: frontend/src/utils/measurement.js
/**
 * 脊柱测量算法 - 与 Next.js 版本完全对齐
 */

const SHOULDER_LEFT = 11
const SHOULDER_RIGHT = 12
const HIP_LEFT = 23
const HIP_RIGHT = 24
const KNEE_LEFT = 25
const KNEE_RIGHT = 26
const EAR_LEFT = 7
const EAR_RIGHT = 8
const ANKLE_LEFT = 27
const ANKLE_RIGHT = 28

const radToDeg = (r) => (r * 180) / Math.PI

const pickLandmark = (list, index) => {
  const lm = list[index]
  if (typeof lm?.x !== 'number' || typeof lm?.y !== 'number') return null
  if (lm.visibility !== undefined && lm.visibility < 0.3) return null
  return lm
}

const dist = (p1, p2) => Math.sqrt((p2.x - p1.x) ** 2 + (p2.y - p1.y) ** 2)

const midpoint = (a, b) => ({
  x: (a.x + b.x) / 2,
  y: (a.y + b.y) / 2,
  visibility: Math.min(a.visibility ?? 1, b.visibility ?? 1)
})

function calcSpinalCurvature(shoulderMid, hipMid, lShoulder, rShoulder, lHip, rHip) {
  const idealMidX = (shoulderMid.x + hipMid.x) / 2
  const torsoLength = dist(shoulderMid, hipMid)
  const maxDeviation = Math.max(
    Math.abs(shoulderMid.x - idealMidX),
    Math.abs(hipMid.x - idealMidX)
  )
  const deviationAngle = radToDeg(Math.atan2(maxDeviation, torsoLength))
  const torsoLateralShift = Math.abs(shoulderMid.x - hipMid.x)
  const lateralShiftAngle = radToDeg(Math.atan2(torsoLateralShift, torsoLength))
  const shoulderAsymmetryAngle = radToDeg(Math.atan2(Math.abs(lShoulder.y - rShoulder.y), 1.0))
  const hipAsymmetryAngle = radToDeg(Math.atan2(Math.abs(lHip.y - rHip.y), 1.0))

  return shoulderAsymmetryAngle * 2.0 + hipAsymmetryAngle * 1.5 + lateralShiftAngle * 1.5 + deviationAngle * 1.0
}

function assessPostureQuality(lShoulder, rShoulder, lHip, rHip, lKnee, rKnee) {
  let confidence = 1.0
  const avgVis = [lShoulder, rShoulder, lHip, rHip]
    .map(lm => lm.visibility ?? 1)
    .reduce((a, b) => a + b, 0) / 4
  if (avgVis < 0.7) confidence *= 0.7

  const widthRatio = Math.abs(lShoulder.x - rShoulder.x) / Math.abs(lHip.x - rHip.x)
  if (widthRatio < 1.0 || widthRatio > 2.0) confidence *= 0.8

  if (lKnee && rKnee) {
    const kneeDistance = Math.abs(lKnee.x - rKnee.x)
    const shoulderDistance = Math.abs(lShoulder.x - rShoulder.x)
    if (kneeDistance > shoulderDistance * 0.8) confidence *= 0.9
  }
  return confidence
}

const SEVERITY_SUMMARY = {
  normal: '脊柱形态整体平衡，当前未见明显异常。建议保持良好站姿和运动习惯，定期复查。',
  mild: '检测到轻度不对称，建议近期关注坐姿、站姿与核心肌群训练，并在 1-2 个月后复测。',
  moderate: '检测到中度异常，建议尽快到医院脊柱外科或康复科进一步评估，必要时结合影像检查确认。',
  severe: '检测到重度异常，建议尽快到医院脊柱外科就诊，并结合 X 光等影像检查进一步明确情况。',
}

const SEVERITY_ORDER = {
  normal: 0,
  mild: 1,
  moderate: 2,
  severe: 3,
}

function pickHigherSeverity(current, next) {
  return (SEVERITY_ORDER[next] ?? -1) > (SEVERITY_ORDER[current] ?? -1) ? next : current
}

function classifyAngleSeverity(angle) {
  const absAngle = Math.abs(angle)
  if (absAngle < 8) return 'normal'
  if (absAngle < 15) return 'mild'
  if (absAngle < 25) return 'moderate'
  return 'severe'
}

function classifyTiltSeverity(angle) {
  const absAngle = Math.abs(angle)
  if (absAngle < 2) return 'normal'
  if (absAngle < 4) return 'mild'
  if (absAngle < 7) return 'moderate'
  return 'severe'
}

function classifyTrunkShiftSeverity(trunkShift) {
  if (trunkShift === null || trunkShift === undefined) return 'normal'
  const absShift = Math.abs(trunkShift)
  if (absShift < 0.05) return 'normal'
  if (absShift < 0.1) return 'mild'
  if (absShift < 0.2) return 'moderate'
  return 'severe'
}

function classifyConfidenceSeverity(confidence) {
  return confidence < 0.6 ? 'mild' : 'normal'
}

function classifySeverity(curvature, shoulderAngle, pelvisAngle, confidence, trunkShift = null, forwardBendSeverity = null) {
  const candidates = [
    classifyAngleSeverity(curvature),
    classifyTiltSeverity(shoulderAngle),
    classifyTiltSeverity(pelvisAngle),
    classifyTrunkShiftSeverity(trunkShift),
    classifyConfidenceSeverity(confidence),
  ]

  const baseSeverity = candidates.reduce((highest, current) => pickHigherSeverity(highest, current), 'normal')

  // 弯腰位作为保守修饰因子，最多提升一级，不可独立产生严重诊断
  if (forwardBendSeverity === 'severe' && baseSeverity === 'normal') return 'mild'
  if (forwardBendSeverity === 'severe' && baseSeverity === 'mild') return 'moderate'
  if (forwardBendSeverity === 'moderate' && baseSeverity === 'normal') return 'mild'
  return baseSeverity
}

export function calculateMetrics(landmarks, imageWidth = 640, imageHeight = 480) {
  if (!landmarks || landmarks.length < 33) {
    throw new Error(`需要33个关键点，但只收到 ${landmarks?.length || 0} 个`)
  }

  const lShoulder = pickLandmark(landmarks, SHOULDER_LEFT)
  const rShoulder = pickLandmark(landmarks, SHOULDER_RIGHT)
  const lHip = pickLandmark(landmarks, HIP_LEFT)
  const rHip = pickLandmark(landmarks, HIP_RIGHT)
  const lKnee = pickLandmark(landmarks, KNEE_LEFT)
  const rKnee = pickLandmark(landmarks, KNEE_RIGHT)

  if (!lShoulder || !rShoulder || !lHip || !rHip) {
    throw new Error('关键关节点不可见，请确保背部完整入镜')
  }

  const shoulderMid = midpoint(lShoulder, rShoulder)
  const hipMid = midpoint(lHip, rHip)

  const postureConfidence = assessPostureQuality(lShoulder, rShoulder, lHip, rHip, lKnee, rKnee)
  const spinalCurvatureDeg = calcSpinalCurvature(shoulderMid, hipMid, lShoulder, rShoulder, lHip, rHip)
  const shoulderHeightDiffPx = Math.abs(lShoulder.y - rShoulder.y) * imageHeight
  const shoulderSlopeDeg = Math.abs(radToDeg(Math.atan2(rShoulder.y - lShoulder.y, rShoulder.x - lShoulder.x)))
  const pelvisTiltDeg = Math.abs(radToDeg(Math.atan2(rHip.y - lHip.y, rHip.x - lHip.x)))

  // 全身力线指标（可选，依赖耳部和踝部关键点）
  const lEar = pickLandmark(landmarks, EAR_LEFT)
  const rEar = pickLandmark(landmarks, EAR_RIGHT)
  const lAnkle = pickLandmark(landmarks, ANKLE_LEFT)
  const rAnkle = pickLandmark(landmarks, ANKLE_RIGHT)

  let trunkShiftNorm = null
  let headTiltDeg = null
  let ankleCompensationRatio = null

  const hipWidth = Math.abs(lHip.x - rHip.x)
  if (hipWidth > 0) {
    trunkShiftNorm = +((shoulderMid.x - hipMid.x) / hipWidth).toFixed(4)
  }

  if (lEar && rEar) {
    headTiltDeg = +radToDeg(Math.atan2(rEar.y - lEar.y, rEar.x - lEar.x)).toFixed(2)
  }

  if (lAnkle && rAnkle) {
    const ankleMidX = (lAnkle.x + rAnkle.x) / 2
    if (hipWidth > 0) {
      ankleCompensationRatio = +((ankleMidX - hipMid.x) / hipWidth).toFixed(4)
    }
  }

  const severity = classifySeverity(
    spinalCurvatureDeg,
    shoulderSlopeDeg,
    pelvisTiltDeg,
    postureConfidence,
    trunkShiftNorm
  )

  return {
    shoulderHeightDiffPx: +shoulderHeightDiffPx.toFixed(1),
    shoulderSlopeDeg: +shoulderSlopeDeg.toFixed(2),
    pelvisTiltDeg: +pelvisTiltDeg.toFixed(2),
    spinalCurvatureDeg: +spinalCurvatureDeg.toFixed(2),
    postureConfidence: +postureConfidence.toFixed(2),
    severity,
    summary: SEVERITY_SUMMARY[severity],
    trunkShiftNorm,
    headTiltDeg,
    ankleCompensationRatio,
  }
}

export function calculateForwardBendMetrics(landmarks) {
  if (!landmarks || landmarks.length < 33) {
    throw new Error(`需要33个关键点，但只收到 ${landmarks?.length || 0} 个`)
  }

  const lShoulder = pickLandmark(landmarks, SHOULDER_LEFT)
  const rShoulder = pickLandmark(landmarks, SHOULDER_RIGHT)
  const lHip = pickLandmark(landmarks, HIP_LEFT)
  const rHip = pickLandmark(landmarks, HIP_RIGHT)

  if (!lShoulder || !rShoulder || !lHip || !rHip) {
    throw new Error('无法检测到肩部和髋部关键点，请确保背部完整入镜')
  }

  // 弯腰位肩部倾斜（有符号，正值=左侧更高）
  const shoulderTiltDeg = +radToDeg(
    Math.atan2(rShoulder.y - lShoulder.y, rShoulder.x - lShoulder.x)
  ).toFixed(2)

  // 弯腰位骨盆倾斜（有符号，同上）
  const pelvisTiltDeg = +radToDeg(
    Math.atan2(rHip.y - lHip.y, rHip.x - lHip.x)
  ).toFixed(2)

  // 肩-骨盆扭转角（绝对值）
  const torsionDeg = +Math.abs(shoulderTiltDeg - pelvisTiltDeg).toFixed(2)

  // 躯干侧移（归一化）
  const shoulderMid = midpoint(lShoulder, rShoulder)
  const hipMid = midpoint(lHip, rHip)
  const hipWidth = Math.abs(lHip.x - rHip.x)
  const trunkShiftNorm = hipWidth > 0
    ? +((shoulderMid.x - hipMid.x) / hipWidth).toFixed(4)
    : null

  // 复合对称评分
  const asymmetryScore = +(
    Math.abs(shoulderTiltDeg) * 0.35 +
    Math.abs(pelvisTiltDeg) * 0.35 +
    torsionDeg * 0.30
  ).toFixed(2)

  // 主导不对称侧
  const shoulderSide = shoulderTiltDeg > 0 ? 'left' : 'right'
  const pelvisSide = pelvisTiltDeg > 0 ? 'left' : 'right'
  const dominantSide = Math.abs(shoulderTiltDeg) >= Math.abs(pelvisTiltDeg)
    ? shoulderSide : pelvisSide

  // 严重度
  const severity = asymmetryScore < 2.0 ? 'none'
    : asymmetryScore < 5.0 ? 'mild'
    : asymmetryScore < 10.0 ? 'moderate'
    : 'severe'

  return {
    asymmetryScore,
    shoulderTiltDeg,
    pelvisTiltDeg,
    torsionDeg,
    trunkShiftNorm,
    dominantSide,
    severity,
  }
}
